docs/decisions/0011_multiple_event_types_per_topic.rst

```diff
@@ -0,0 +1,74 @@
+.. _Putting Several Event Types in the Same Topic – Revisited: https://www.confluent.io/blog/multiple-event-types-in-the-same-kafka-topic/
+
+Decision
+********
+We will update the code for event serializers, which are responsible for registering and validating schemas against the schema registry, to use the TopicRecordName strategy.
+
+TopicRecordName Strategy
+========================
+The Confluent Schema Registry stores schemas according to one of three strategies, referring to the identifiers under which the schemas are stored. When an event is serialized using an AvroSerializer, the serializer checks the schema against the registry, looking up the existing schema using the configured strategy. If the new schema is not allowable under the schema evolution rules, serialization will fail.
+
+The default strategy the TopicName strategy. Under this strategy, the serializer will determine the existing schema by looking for a schema registered as ``<topic-name>-value`` (``<topic-name>-key`` when serializing the key), the topic name having been passed in the SerializationContext object. Therefore, if the serializer is given a new event type with the same topic, unless the new event type schema happens to be a valid evolution of the existing one, it will fail.
+
+Using the TopicRecordName strategy, schemas are registered by topic and record name. On serializing an event, the serializer looks for a schema registered as ``<topic-name>-<record-name>``.
+For example, given a schema like::
+
+    {
+        "doc": "Avro Event Format for CloudEvents created with openedx_events/schema",
+        "fields": [{"name": "str_field", "type": "string"}]
+        "name": "CloudEvent",
+        "namespace": "my.event.type",
+        "type": "record"
+    }
+
+and the topic ``my-topic``, the serializer will look for a schema registered as ``my-topic-my.event.type.CloudEvent`` and use that to determine if the schema being passed is valid. This means that as long as the namespace is different for the schemas for different event types, the serializer will not error when adding new event types. It will still maintain evolution rules within an event type, however, which is desirable behavior.
+
+Consequences
+************
+* The record names for each event type will need to be distinct from each other. Originally, they were all just "CloudEvent." We needed to add namespaces to ensure that they were unique. Note that this will be necessary regardless of which solution we choose.
+* Any event type will be able to be written to any topic
+* Schema evolution rules will be enforced for event types but not for topics.
+* The event-bus-kafka consumer will need to be updated to determine the signal from the message headers rather than taking a signal as a passed argument in the management command. This will also be necessary regardless of which solution we choose.
+
```

Comment:
Somewhere could we note that in the future we might be able to detect if a service is only sending one event type for topic for all topics, and to not configure in the same way in this case?

Reply:
I'm not sure that's true. See the consequences of using a rollout toggle. We pretty much need this to be all-or-nothing across services.

Reply:
Maybe we can have a chat, simply because I don't follow it. Only a single service should ever produce to a topic. I don't think we have an ADR around that, but that is a pattern that seems to be generally accepted. There might be several instances (e.g. EC2 instances, k8s containers) of the same service, but I don't think that is what you are referring to...

As I was writing this, I had a thought that _may_ be what you were trying to say. The above constraint is possibly about an event type, and not a topic. So, one question is whether this ADR should declare that all event types being produced to a shared topic must be sourced from the same service. I say yes.

Does this have anything to do with the point you were trying to make, or do we also need an entirely separate discussion as well?

Reply:
I was thinking about multiple services producing to the same topic. If we are comfortable saying that is not allowed and if you do that you could end up with weird schema problems, that's fine. I would definitely call that out in the ADR though.

Reply:
[fixed] Added back the SettingsToggle, and noted that this requires one producer per topic